server/container_create.go

```diff
@@ -82,18 +82,18 @@ func addOCIBindMounts(mountLabel string, containerConfig *pb.ContainerConfig, sp
 			return nil, nil, fmt.Errorf("Mount.ContainerPath is empty")
 		}
 
-		src := mount.HostPath
-		if src == "" {
+		if mount.HostPath == "" {
 			return nil, nil, fmt.Errorf("Mount.HostPath is empty")
 		}
+		src := filepath.Join(bindMountPrefix, mount.HostPath)
 
-		if _, err := os.Stat(src); err != nil && os.IsNotExist(err) {
+		if _, err := os.Lstat(src); err != nil && os.IsNotExist(err) {
 			if err1 := os.MkdirAll(src, 0644); err1 != nil {
```

Comment:
The `Stat` → `Lstat` change above makes some sense to me.  But if there's a symlink at `src` where one of the paths in its resolution chain does not exist, I think we need to create that file at the end of the resolution chain.  As it stands here, we're attempting to create `src`, which will fail because of the existing symlink at that location.  Anyhow, perhaps the `Stat` → `Lstat` change is orthogonal enough to spin off into its own PR?

Reply:
Without this change the PR won't work at all with a symlink.

Does `os.MkdirAll` follow symlinks?  I'd expect it to fail immediately with a dangling symlink.

Alternatively, should we try to resolve the symlink before the Stat/Mkdir?

Reply:
> Does `os.MkdirAll` follow symlinks? I'd expect it to fail immediately with a dangling symlink.

It fails when the target is a dangling symlink, as you expected.

> Alternatively, should we try to resolve the symlink before the Stat/Mkdir?

This seems like the most robust approach to me, although it's not clear to me how much work we want to put into helping the caller create bind-mount source directories to exist.  If we want to support the “caller asked me to bind mount from a dangling symlink” use case, here's some test code to draw from:

```go
package main

import (
	"errors"
	"fmt"
	"os"
	"syscall"
)

var maxDepthErr = errors.New("maximum depth exceeded")

func main() {
	os.Remove("a") // ignore errors
	os.Remove("b") // ignore errors
	os.Remove("c") // ignore errors
	err := os.MkdirAll("d", 0777)
	if err != nil {
		fmt.Printf("%+v\n", err)
		os.Exit(1)
	}
	err = os.Symlink("b", "a") // a -> b
	if err != nil {
		fmt.Printf("%+v\n", err)
		os.Exit(1)
	}
	err = os.Symlink("c", "b") // b -> c
	if err != nil {
		fmt.Printf("%+v\n", err)
		os.Exit(1)
	}
	info, err := os.Lstat("a")
	if err != nil {
		fmt.Printf("%+v\n", err)
		os.Exit(1)
	}
	fmt.Printf("lstat is symlink? %t\n", info.Mode()&os.ModeSymlink != 0)
	info, err = os.Stat("a")
	if err == nil {
		fmt.Println("unexpected stat success")
		os.Exit(1)
	}
	err = syscall.Mount("a", "d", "", syscall.MS_BIND, "")
	if err == nil {
		fmt.Println("unexpected mount success")
		os.Exit(1)
	}
	if !os.IsNotExist(err) {
		fmt.Printf("%+v\n", err)
		os.Exit(1)
	}
	err = setupSource("a", 0777, 0, 10)
	if err != nil {
		fmt.Printf("%+v\n", err)
		os.Exit(1)
	}
	err = syscall.Mount("a", "d", "", syscall.MS_BIND, "")
	if err != nil {
		fmt.Printf("%+v\n", err)
		os.Exit(1)
	}
	err = syscall.Unmount("d", 0)
	if err != nil {
		fmt.Printf("%+v\n", err)
		os.Exit(1)
	}
}

func setupSource(src string, perm os.FileMode, depth int, maxDepth int) (err error) {
	if depth > maxDepth {
		return maxDepthErr
	}
	info, err := os.Lstat(src)
	if err == nil {
		if info.IsDir() {
			return nil
		}
		if info.Mode()&os.ModeSymlink != 0 {
			target, err := os.Readlink(src)
			if err != nil {
				return err
			}
			return setupSource(target, perm, depth+1, maxDepth)
		}
		return fmt.Errorf("%q resolves to a non-directory", src)
	}
	if !os.IsNotExist(err) {
		return err
	}
	return os.MkdirAll(src, perm)
}
```

Reply:
is this a blocking issue or can be addressed later in a separate PR?

Reply:
> is this a blocking issue or can be addressed later in a separate PR?

I'm not a maintainer, so it's not my call.  But [I don't see why the whole `Stat` → `Lstat` can't be spun off into a separate PR][1].  Is there something about the rest of this PR that exposes the dangling-symlink issue?

[1]: https://github.com/kubernetes-incubator/cri-o/pull/1485#discussion_r179241882

Reply:
`os.Stat` solves the symlink in the wrong context when the prefix is specified, so it cannot be used.
In the case of not using a prefix, is it different that what we are doing now?  `MkdirAll` would still fail if `src` is a symlink.

Reply:
> `os.Stat` solves the symlink in the wrong context when the prefix is specified, so it cannot be used.

Ah, right.  So I'm good with your `Stat` → `Lstat` as it stands, and can follow up with retargeting the `MkdirAll` after this PR lands.